### ISSUE TITLE
3.0: Callback for entering and exiting of one proximity zone

### ISSUE THREAD

**Issue by egmontr** [AltBeacon/android-beacon-library#257]
Remember the entering and exiting of one proximity zone (range) and report it to the app.

**Comment by davidgyoung** [comment]
I agree this would be a useful enhancement.  There are two challenges in coming up with a good implementation.

1.  Keep entry/exits from happening repeatedly when on the edge of the proximity zone.

The problem is that the estimated distance fluctuates, so if the mobile device is near the boundary, it would cause repeated entry/exit events.  One possibility is to have a "threshold" parameter, which defaults to a value that is 100% of the proximity distance.  So it might work like this:

     ProximityFilter proximityFilter = new ProximityFilter(5, 10);  // 5 meter distance filter with a 10 meter exit filter.  Same as calling new ProximityFilter(5);
     region.setProximityFilter(proximityFilter);

Monitoring on the region above would only trigger an entry when the estimated distance is < 5 meters, and an exit would then only trigger if the distance > 10 meters.

2. On Android 5+, battery savings are achieved by returning background detection events only when a bluetooth hardware filter is triggered.  The hardware filter is not smart enough to handle distance estimates, so this would have to be done in software.  Under the current implementation, the software processing of packets only happens for the first 10 seconds after the beacon is first detected (unless the app is brought to the foreground).  This would have to change, otherwise the mobile device would not trigger an entry event if a proximity-based filter were in place and the mobile device did not come within the filter distance in the first 10 seconds after detection.
